README.md
# PostCSS Variable Compress ![npm](https://img.shields.io/npm/dy/postcss-variable-compress) ![GitHub branch checks state](https://img.shields.io/github/checks-status/navanshu/postcss-variable-compress/master)

[postcss-variable-compress] is a [PostCSS] plugin minifies variable names and saves space. Even if you have 1295 css variables still they will not exceed by two characters. It will transform css variable without breaking your stylesheet. Get if from [NPM].

If you want it not modify some css variables, then pass them `--{variable-name}` as an array to the plugin.

[postcss]: https://github.com/postcss/postcss
[postcss-variable-compress]: https://github.com/navanshu/postcss-variable-compress
[npm]: https://www.npmjs.com/package/postcss-variable-compress

### If you are looking for a plugin that can work on separate files go to import splitFiles.js, works the same but it doesn't resets the variables.

```css
:root {
  --first-color: #16f;
  --second-color: #ff7;
}

#firstParagraph {
  background-color: var(--first-color);
  color: var(--second-color);
}

#container {
  --first-color: #290;
}
```

```css
:root {
  --0: #16f;
  --1: #ff7;
}

#firstParagraph {
  background-color: var(--0);
  color: var(--1);
}

#container {
  --0: #290;
}
```

## Usage

**Step 1:** Install plugin:

```sh
npm install --save-dev postcss postcss-variable-compress
```

**Step 2:** Check you project for existed PostCSS config: `postcss.config.js`
in the project root, `"postcss"` section in `package.json`
or `postcss` in bundle config.

If you do not use PostCSS, add it according to [official docs]
and set this plugin in settings.

**Step 3:** Add the plugin at the end of the plugins list:

```diff
module.exports = {
  plugins: [
    require('cssnano'),
+   require('postcss-variable-compress')
  ]
}
```

## Api

```javascript

module.exports = {
  plugins: [
    require('cssnano'),
    require('postcss-variable-compress')(
      // pass in css variables to avoid
      'light', 'dark', 'font', 'vh', 'r' // unprefixed
      // or
      '--height', '--font' // prefixed
      // or pass in as many function as your want
      // they take single param which is a the name of css variable
      // you can do checks on it and
      // return true if you want it to be skipped
      // for example
      (name) => name.includes('skip') // name prefixed css variable example --height
      // avoid regex if you can they are bad
    )
  ]
}

```

[official docs]: https://github.com/postcss/postcss#usage


CHANGELOG.md
# Change Log

# V2.0.0

- Exposed the function fixing problems with typescript in regards to [Pull Request by Kaciras](https://github.com/navanshu/postcss-variable-compress/pull/38)

# V0.2.6

- Updating issues
- Update postcss

# V0.2.5

- Coverage 100%.

# v0.2.0

- Same api can be passed with function so you can check if a variable has to be ignored

# v0.1.3

- Fixed the scoping issue in the variable
- Trying to circumvent the Regex. Can't remove it yet but have some checks before regex in place.
- Plugin notifies postcss about if a Declaration has bee processed.

# v0.1.2

- Fixing readme

# v0.1.1

- Updated readme

# v0.1.0

- Initial release


index.test.js
const postcss = require('postcss');

const variableCompress = require('./index');

async function run (input, output, opts) {
  let result = await postcss(
    [variableCompress(opts)]
  ).process(input, { from: undefined });

  expect(result.css).toEqual(output);
  expect(result.warnings()).toHaveLength(0);

}


it('Shorten css variables', async () => {
  await run(
    `:root {
--first-color: #16f;
--second-color: #ff7;
--2: #000;
}

#firstParagraph {
  background-color: var(--first-color);
  color: var(--second-color);
}

#secondParagraph {
  background-color: var(--second-color);
  color: var(--first-color);
}

#container {
  --first-color: #290;
}

#thirdParagraph {
  background-color: var(--first-color);
  color: var(--second-color);
}

.section-title {
  color: var(--primary-color, var(--black, #222));
}

code {
  --5: #555;
}`,
    `:root {
--0: #16f;
--1: #ff7;
--2: #000;
}

#firstParagraph {
  background-color: var(--0);
  color: var(--1);
}

#secondParagraph {
  background-color: var(--1);
  color: var(--0);
}

#container {
  --0: #290;
}

#thirdParagraph {
  background-color: var(--0);
  color: var(--1);
}

.section-title {
  color: var(--primary-color, var(--3, #222));
}

code {
  --5: #555;
}`,
    [
      '--primary-color',
      '2',
      (e) => e.includes('special'),
      (e) => e === '--5',
    ]
  );
});

it('Shorten known css variables', async () => {
  const knownClassNames = {
    '--first-color': '--a',
    '--second-color': '--b',
  }

  await run(
    `:root {
--first-color: #16f;
--second-color: #ff7;
--2: #000;
}

#firstParagraph {
  background-color: var(--first-color);
  color: var(--second-color);
}

#secondParagraph {
  background-color: var(--second-color);
  color: var(--first-color);
}

#container {
  --first-color: #290;
}

#thirdParagraph {
  background-color: var(--first-color);
  color: var(--second-color);
}

.section-title {
  color: var(--primary-color, var(--black, #222));
}

code {
  --5: #555;
}`,
    `:root {
--a: #16f;
--b: #ff7;
--2: #000;
}

#firstParagraph {
  background-color: var(--a);
  color: var(--b);
}

#secondParagraph {
  background-color: var(--b);
  color: var(--a);
}

#container {
  --a: #290;
}

#thirdParagraph {
  background-color: var(--a);
  color: var(--b);
}

.section-title {
  color: var(--primary-color, var(--0, #222));
}

code {
  --5: #555;
}`,
    [
      '--primary-color',
      '2',
      (e) => e.includes('special'),
      (e) => e === '--5',
      knownClassNames,
    ]
  );
});

it('Support reloading. Now the plugin will reset mapped variables', async () => {
  await run(`:root{--first-color: #16f;--second-color: #ff7;}`, `:root{--0: #16f;--1: #ff7;}`, []);
  await run(`:root{--second-color: #ff7;--first-color: #16f;}`, `:root{--0: #ff7;--1: #16f;}`, []);
});

it('Base array check or no array', async () => {
  await run(`:root{}`, `:root{}`);
});


splitFiles.test.js
const postcss = require('postcss');

const variableCompress = require('./splitFiles');

async function run (input, output, opts) {
  let result = await postcss(
    [variableCompress(opts)]
  ).process(input, { from: undefined });

  expect(result.css).toEqual(output);
  expect(result.warnings()).toHaveLength(0);

}


it('Shorten css variables', async () => {
  await run(
    `:root {
--first-color: #16f;
--second-color: #ff7;
--2: #000;
}

#firstParagraph {
  background-color: var(--first-color);
  color: var(--second-color);
}

#secondParagraph {
  background-color: var(--second-color);
  color: var(--first-color);
}

#container {
  --first-color: #290;
}

#thirdParagraph {
  background-color: var(--first-color);
  color: var(--second-color);
}

.section-title {
  color: var(--primary-color, var(--black, #222));
}

code {
  --5: #555;
}`,
    `:root {
--0: #16f;
--1: #ff7;
--2: #000;
}

#firstParagraph {
  background-color: var(--0);
  color: var(--1);
}

#secondParagraph {
  background-color: var(--1);
  color: var(--0);
}

#container {
  --0: #290;
}

#thirdParagraph {
  background-color: var(--0);
  color: var(--1);
}

.section-title {
  color: var(--primary-color, var(--3, #222));
}

code {
  --5: #555;
}`,
    [
      '--primary-color',
      '2',
      (e) => e.includes('special'),
      (e) => e === '--5'
    ]
  );
});

it('No reloading', async () => {
  await run(`:root{--first-color: #16f;--second-color: #ff7;}`, `:root{--0: #16f;--1: #ff7;}`, []);
  await run(`:root{--second-color: #ff7;--first-color: #16f;}`, `:root{--1: #ff7;--0: #16f;}`, []);
});


it('Base array check or no array', async () => {
  await run(`:root{}`, `:root{}`);
});


index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const postcssPlugin = "postcss-variable-compress";
let processed = Symbol("processed");
let renamedVariables = [];
let cssVariables = -1;
let pureSkips = [];
let scriptBasedSkips = [];
let cssVariablesMap = new Map();
function scriptCheck(val) {
    const should = scriptBasedSkips.findIndex((E) => E(val));
    return should > -1;
}
function shouldRun(val) {
    let should = true;
    if (renamedVariables.indexOf(val) > -1) {
        should = false;
    }
    if (should && scriptCheck(val)) {
        should = false;
    }
    return should;
}
function increase() {
    cssVariables = cssVariables + 1;
    let temp = cssVariables.toString(36);
    pureSkips.forEach((E) => {
        if (E === temp) {
            temp = cssVariables.toString(36);
            cssVariables = cssVariables + 1;
        }
    });
}
function replacer(match) {
    if (!shouldRun(match))
        return match;
    let exist = cssVariablesMap.get(match);
    if (!exist) {
        exist = "--" + cssVariables.toString(36);
        increase();
        cssVariablesMap.set(match, exist);
        renamedVariables.push(exist);
    }
    return exist;
}
function map(j) {
    let prop = j.prop;
    if (prop && j.variable && shouldRun(prop)) {
        let old = prop;
        let exist = cssVariablesMap.get(old);
        if (!exist) {
            exist = "--" + cssVariables.toString(36);
            increase();
            cssVariablesMap.set(old, exist);
            renamedVariables.push(exist);
        }
        prop = exist;
        j.prop = prop;
    }
    let value = j.value;
    if (value && value.includes("var(--") && value.length <= 1000) {
        value = value.replace(/--[\w-_]{1,1000}/g, replacer);
        j.value = value;
    }
    // @ts-ignore
    j[processed] = true;
}
function variableCompress(opts) {
    processed = Symbol("processed");
    renamedVariables = [];
    cssVariables = -1;
    pureSkips = [];
    scriptBasedSkips = [];
    cssVariablesMap = new Map();
    opts === null || opts === void 0 ? void 0 : opts.forEach((E) => {
        if (typeof E === "string") {
            let name = E;
            let cssName = E;
            if (E.slice(0, 2) === "--") {
                name = E.slice(2);
            }
            else {
                cssName = "--" + E;
            }
            pureSkips.push(name);
            cssVariablesMap.set(cssName, cssName);
        }
        else
            scriptBasedSkips.push(E);
    });
    increase();
    return {
        postcssPlugin,
        Declaration: {
            "*": map,
        },
    };
}
module.exports = variableCompress;
module.exports.postcss = true;
exports.default = variableCompress;


splitFiles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const postcssPlugin = "postcss-variable-compress";
let processed = Symbol("processed");
let renamedVariables = [];
let cssVariables = -1;
let pureSkips = [];
let scriptBasedSkips = [];
let cssVariablesMap = new Map();
function scriptCheck(val) {
    const should = scriptBasedSkips.findIndex((E) => E(val));
    return should > -1;
}
function shouldRun(val) {
    let should = true;
    if (renamedVariables.indexOf(val) > -1) {
        should = false;
    }
    if (should && scriptCheck(val)) {
        should = false;
    }
    return should;
}
function increase() {
    cssVariables = cssVariables + 1;
    let temp = cssVariables.toString(36);
    pureSkips.forEach((E) => {
        if (E === temp) {
            temp = cssVariables.toString(36);
            cssVariables = cssVariables + 1;
        }
    });
}
function replacer(match) {
    if (!shouldRun(match))
        return match;
    let exist = cssVariablesMap.get(match);
    if (!exist) {
        exist = "--" + cssVariables.toString(36);
        increase();
        cssVariablesMap.set(match, exist);
        renamedVariables.push(exist);
    }
    return exist;
}
function map(j) {
    let prop = j.prop;
    if (prop && j.variable && shouldRun(prop)) {
        let old = prop;
        let exist = cssVariablesMap.get(old);
        if (!exist) {
            exist = "--" + cssVariables.toString(36);
            increase();
            cssVariablesMap.set(old, exist);
            renamedVariables.push(exist);
        }
        prop = exist;
        j.prop = prop;
    }
    let value = j.value;
    if (value && value.includes("var(--") && value.length <= 1000) {
        value = value.replace(/--[\w-_]{1,1000}/g, replacer);
        j.value = value;
    }
    // @ts-ignore
    j[processed] = true;
}
function variableCompressSplitFiles(opts) {
    processed = Symbol("processed");
    opts === null || opts === void 0 ? void 0 : opts.forEach((E) => {
        if (typeof E === "string") {
            let name = E;
            let cssName = E;
            if (E.slice(0, 2) === "--") {
                name = E.slice(2);
            }
            else {
                cssName = "--" + E;
            }
            pureSkips.push(name);
            cssVariablesMap.set(cssName, cssName);
        }
        else
            scriptBasedSkips.push(E);
    });
    increase();
    return {
        postcssPlugin,
        Declaration: {
            "*": map,
        },
    };
}
module.exports = variableCompressSplitFiles;
module.exports.postcss = true;
exports.default = variableCompressSplitFiles;


splitFiles.ts
export namespace variableCompressSplitFiles {
  export type skip = (variableName: string) => boolean | undefined;
  export type parameters = skip | string;
}

const postcssPlugin = "postcss-variable-compress";

let processed = Symbol("processed");
let renamedVariables: string[] = [];
let cssVariables = -1;
let pureSkips: string[] = [];
let scriptBasedSkips: variableCompressSplitFiles.skip[] = [];
let cssVariablesMap = new Map();

function scriptCheck(val: string) {
  const should = scriptBasedSkips.findIndex((E) => E(val));
  return should > -1;
}

function shouldRun(val: string) {
  let should = true;

  if (renamedVariables.indexOf(val) > -1) {
    should = false;
  }

  if (should && scriptCheck(val)) {
    should = false;
  }

  return should;
}

function increase() {
  cssVariables = cssVariables + 1;

  let temp = cssVariables.toString(36);

  pureSkips.forEach((E) => {
    if (E === temp) {
      temp = cssVariables.toString(36);
      cssVariables = cssVariables + 1;
    }
  });
}

function replacer(match: string) {
  if (!shouldRun(match)) return match;
  let exist = cssVariablesMap.get(match);

  if (!exist) {
    exist = "--" + cssVariables.toString(36);
    increase();
    cssVariablesMap.set(match, exist);
    renamedVariables.push(exist);
  }

  return exist;
}

function map(j: import("postcss").Declaration) {
  let prop = j.prop;
  if (prop && j.variable && shouldRun(prop)) {
    let old = prop;
    let exist = cssVariablesMap.get(old);
    if (!exist) {
      exist = "--" + cssVariables.toString(36);
      increase();
      cssVariablesMap.set(old, exist);
      renamedVariables.push(exist);
    }
    prop = exist;
    j.prop = prop;
  }

  let value = j.value;
  if (value && value.includes("var(--") && value.length <= 1000) {
    value = value.replace(/--[\w-_]{1,1000}/g, replacer);
    j.value = value;
  }

  // @ts-ignore
  j[processed] = true;
}

function variableCompressSplitFiles(
  opts?: variableCompressSplitFiles.parameters[]
) {
  processed = Symbol("processed");

  opts?.forEach((E) => {
    if (typeof E === "string") {
      let name = E;
      let cssName = E;

      if (E.slice(0, 2) === "--") {
        name = E.slice(2);
      } else {
        cssName = "--" + E;
      }

      pureSkips.push(name);
      cssVariablesMap.set(cssName, cssName);
    } else scriptBasedSkips.push(E);
  });

  increase();

  return {
    postcssPlugin,
    Declaration: {
      "*": map,
    },
  };
}

module.exports = variableCompressSplitFiles;
module.exports.postcss = true;
export default variableCompressSplitFiles;


index.ts
export namespace variableCompress {
  export type skip = (variableName: string) => boolean | undefined;

  export type parameters = skip | string;
}
const postcssPlugin = "postcss-variable-compress";

let processed = Symbol("processed");
let renamedVariables: string[] = [];
let cssVariables = -1;
let pureSkips: string[] = [];
let scriptBasedSkips: variableCompress.skip[] = [];
let cssVariablesMap = new Map();

function scriptCheck(val: string) {
  const should = scriptBasedSkips.findIndex((E) => E(val));
  return should > -1;
}

function shouldRun(val: string) {
  let should = true;

  if (renamedVariables.indexOf(val) > -1) {
    should = false;
  }

  if (should && scriptCheck(val)) {
    should = false;
  }

  return should;
}

function increase() {
  cssVariables = cssVariables + 1;

  let temp = cssVariables.toString(36);

  pureSkips.forEach((E) => {
    if (E === temp) {
      temp = cssVariables.toString(36);
      cssVariables = cssVariables + 1;
    }
  });
}

function replacer(match: string) {
  if (!shouldRun(match)) return match;
  let exist = cssVariablesMap.get(match);

  if (!exist) {
    exist = "--" + cssVariables.toString(36);
    increase();
    cssVariablesMap.set(match, exist);
    renamedVariables.push(exist);
  }

  return exist;
}

function map(j: import("postcss").Declaration) {
  let prop = j.prop;
  if (prop && j.variable && shouldRun(prop)) {
    let old = prop;
    let exist = cssVariablesMap.get(old);
    if (!exist) {
      exist = "--" + cssVariables.toString(36);
      increase();
      cssVariablesMap.set(old, exist);
      renamedVariables.push(exist);
    }
    prop = exist;
    j.prop = prop;
  }

  let value = j.value;
  if (value && value.includes("var(--") && value.length <= 1000) {
    value = value.replace(/--[\w-_]{1,1000}/g, replacer);
    j.value = value;
  }

  // @ts-ignore
  j[processed] = true;
}

function variableCompress(opts?: variableCompress.parameters[]) {
  processed = Symbol("processed");

  renamedVariables = [];
  cssVariables = -1;
  pureSkips = [];
  scriptBasedSkips = [];
  cssVariablesMap = new Map();

  opts?.forEach((E) => {
    if (typeof E === "string") {
      let name = E;
      let cssName = E;

      if (E.slice(0, 2) === "--") {
        name = E.slice(2);
      } else {
        cssName = "--" + E;
      }

      pureSkips.push(name);
      cssVariablesMap.set(cssName, cssName);
    } else scriptBasedSkips.push(E);
  });

  increase();

  return {
    postcssPlugin,
    Declaration: {
      "*": map,
    },
  };
}

module.exports = variableCompress;
module.exports.postcss = true;
export default variableCompress;
